File: frontend/src/api/billing.ts
import apiClient from './client'
import type { PaymentOrder } from './wallet'

export type BillingDurationUnit = 'day' | 'month' | 'year' | 'custom'
export type BillingPurchaseLimitScope = 'active_period' | 'lifetime' | 'unlimited'
export type WalletCreditBucket = 'recharge' | 'gift'

export interface EpayChannelConfig {
  channel: string
  display_name: string
  fee_rate?: number
}

export interface EpayGatewayConfig {
  provider: PaymentGatewayProvider
  enabled: boolean
  endpoint_url?: string | null
  callback_base_url?: string | null
  merchant_id?: string | null
  has_secret: boolean
  has_secret_keys?: string[]
  pay_currency?: string | null
  usd_exchange_rate?: number | null
  min_recharge_usd?: number | null
  channels?: EpayChannelConfig[]
  refund_enabled?: boolean
  allow_user_refund?: boolean
  config?: Record<string, unknown>
  created_at?: number | null
  updated_at?: number | null
}

export type PaymentGatewayProvider = 'epay' | 'alipay' | 'wxpay' | 'stripe'

export interface UpdateEpayGatewayConfigRequest {
  enabled: boolean
  endpoint_url?: string
  callback_base_url?: string | null
  merchant_id?: string
  merchant_key?: string
  pay_currency: string
  usd_exchange_rate: number
  min_recharge_usd: number
  channels: EpayChannelConfig[]
  refund_enabled?: boolean
  allow_user_refund?: boolean
  config?: Record<string, unknown>
  secrets?: Record<string, string>
}

export interface GatewayTestResponse {
  ok: boolean
  provider: string
}

export interface WalletCreditEntitlement {
  type: 'wallet_credit'
  amount_usd: number
  balance_bucket?: WalletCreditBucket
}

export interface DailyQuotaEntitlement {
  type: 'daily_quota'
  daily_quota_usd: number
  reset_timezone?: string
  carry_over?: boolean
  allow_wallet_overage?: boolean
}

export interface MembershipGroupEntitlement {
  type: 'membership_group'
  grant_user_groups: string[]
}

export type BillingEntitlement =
  | WalletCreditEntitlement
  | DailyQuotaEntitlement
  | MembershipGroupEntitlement

export interface BillingPlan {
  id: string
  title: string
  description?: string | null
  price_amount: number
  price_currency: string
  duration_unit: BillingDurationUnit
  duration_value: number
  enabled: boolean
  sort_order: number
  max_active_per_user: number
  purchase_limit_scope: BillingPurchaseLimitScope
  entitlements: BillingEntitlement[]
  created_at?: number | null
  updated_at?: number | null
}

export interface BillingPlanWriteRequest {
  title: string
  description?: string | null
  price_amount: number
  price_currency: string
  duration_unit: BillingDurationUnit
  duration_value: number
  enabled: boolean
  sort_order: number
  max_active_per_user: number
  purchase_limit_scope: BillingPurchaseLimitScope
  entitlements: BillingEntitlement[]
}

export interface BillingPlanListResponse {
  items: BillingPlan[]
  total: number
}

export interface BillingCheckoutRequest {
  payment_method?: string
  payment_provider?: string
  payment_channel?: string
}

export interface BillingCheckoutResponse {
  order: PaymentOrder & {
    order_kind?: string
    product_id?: string | null
    product?: BillingPlan | null
  }
  payment_instructions: Record<string, unknown>
  reused_pending_order?: boolean
}

export interface UserPlanEntitlement {
  id: string
  user_id: string
  plan_id: string
  payment_order_id: string
  status: string
  starts_at: string | null
  expires_at: string | null
  entitlements: BillingEntitlement[]
  active?: boolean
  created_at?: string | null
  updated_at?: string | null
}

export interface UserPlanEntitlementsResponse {
  items: UserPlanEntitlement[]
  total: number
}

function normalizeChannels(channels: EpayGatewayConfig['channels']): EpayChannelConfig[] {
  return Array.isArray(channels)
    ? channels
      .map((item) => {
        const raw = item as EpayChannelConfig & { type?: string }
        const channel = String(raw.channel || raw.type || '').trim()
        const feeRate = Number(raw.fee_rate ?? 0)
        return {
          channel,
          display_name: String(raw.display_name || channel).trim(),
          fee_rate: Number.isFinite(feeRate) && feeRate >= 0 ? feeRate : 0,
        }
      })
      .filter((item) => item.channel && item.display_name)
    : []
}

function normalizeGatewayConfig(config: EpayGatewayConfig): EpayGatewayConfig {
  const refundEnabled = Boolean(config.refund_enabled)
  return {
    provider: config.provider || 'epay',
    enabled: Boolean(config.enabled),
    endpoint_url: config.endpoint_url ?? '',
    callback_base_url: config.callback_base_url ?? '',
    merchant_id: config.merchant_id ?? '',
    has_secret: Boolean(config.has_secret),
    has_secret_keys: Array.isArray(config.has_secret_keys) ? config.has_secret_keys : [],
    pay_currency: config.pay_currency ?? 'CNY',
    usd_exchange_rate: Number(config.usd_exchange_rate ?? 7.2),
    min_recharge_usd: Number(config.min_recharge_usd ?? 1),
    channels: normalizeChannels(config.channels),
    refund_enabled: refundEnabled,
    allow_user_refund: refundEnabled && Boolean(config.allow_user_refund),
    config: config.config && typeof config.config === 'object' ? config.config : {},
    created_at: config.created_at ?? null,
    updated_at: config.updated_at ?? null,
  }
}

export const epayGatewayApi = {
  async get(provider: PaymentGatewayProvider = 'epay'): Promise<EpayGatewayConfig> {
    const response = await apiClient.get<EpayGatewayConfig>(`/api/admin/payments/gateways/${provider}`)
    return normalizeGatewayConfig(response.data)
  },

  async update(
    payload: UpdateEpayGatewayConfigRequest,
    provider: PaymentGatewayProvider = 'epay'
  ): Promise<EpayGatewayConfig> {
    const request: UpdateEpayGatewayConfigRequest = {
      ...payload,
      channels: normalizeChannels(payload.channels),
      refund_enabled: Boolean(payload.refund_enabled),
      allow_user_refund: Boolean(payload.refund_enabled && payload.allow_user_refund),
    }
    const response = await apiClient.put<EpayGatewayConfig>(`/api/admin/payments/gateways/${provider}`, request)
    return normalizeGatewayConfig(response.data)
  },

  async test(provider: PaymentGatewayProvider = 'epay'): Promise<GatewayTestResponse> {
    const response = await apiClient.post<GatewayTestResponse>(`/api/admin/payments/gateways/${provider}/test`, {})
    return response.data
  },
}

export const adminBillingPlansApi = {
  async list(): Promise<BillingPlanListResponse> {
    const response = await apiClient.get<BillingPlanListResponse>('/api/admin/billing/plans')
    return response.data
  },

  async create(payload: BillingPlanWriteRequest): Promise<BillingPlan> {
    const response = await apiClient.post<BillingPlan>('/api/admin/billing/plans', payload)
    return response.data
  },

  async update(planId: string, payload: BillingPlanWriteRequest): Promise<BillingPlan> {
    const response = await apiClient.put<BillingPlan>(`/api/admin/billing/plans/${planId}`, payload)
    return response.data
  },

  async setStatus(planId: string, enabled: boolean): Promise<BillingPlan> {
    const response = await apiClient.patch<BillingPlan>(`/api/admin/billing/plans/${planId}/status`, { enabled })
    return response.data
  },

  async delete(planId: string): Promise<void> {
    await apiClient.delete(`/api/admin/billing/plans/${planId}`)
  },
}

export const billingApi = {
  async listPlans(): Promise<BillingPlanListResponse> {
    const response = await apiClient.get<BillingPlanListResponse>('/api/billing/plans')
    return response.data
  },

  async checkout(planId: string, payload: BillingCheckoutRequest): Promise<BillingCheckoutResponse> {
    const response = await apiClient.post<BillingCheckoutResponse>(
      `/api/billing/plans/${planId}/checkout`,
      payload
    )
    return response.data
  },

  async listEntitlements(): Promise<UserPlanEntitlementsResponse> {
    const response = await apiClient.get<UserPlanEntitlementsResponse>('/api/billing/entitlements')
    return response.data
  },
}

File: apps/aether-gateway/src/handlers/public/support/billing.rs
use super::support_payment::payment_epay::{
    build_epay_checkout_url, epay_callback_base_url, load_epay_config, resolve_epay_channel,
    EpayCheckoutInput,
};
use super::{
    build_auth_error_response, build_auth_json_response, resolve_authenticated_local_user,
    sanitize_wallet_gateway_response, unix_secs_to_rfc3339, AppState, GatewayPublicRequestContext,
};
use crate::handlers::shared::{
    create_alipay_direct_checkout, create_stripe_direct_checkout, create_wxpay_direct_checkout,
    direct_payment_client_ip, DirectPaymentCheckoutInput,
};
use axum::{
    body::{Body, Bytes},
    http,
    response::{IntoResponse, Response},
    Json,
};
use chrono::Utc;
use serde::Deserialize;
use serde_json::{json, Value};
use uuid::Uuid;

const BILLING_STORAGE_UNAVAILABLE_DETAIL: &str = "套餐后端暂不可用";

#[derive(Debug, Deserialize, Default)]
struct BillingPlanCheckoutRequest {
    #[serde(default)]
    payment_method: Option<String>,
    #[serde(default)]
    payment_provider: Option<String>,
    #[serde(default)]
    payment_channel: Option<String>,
}

#[derive(Debug, Clone)]
struct NormalizedBillingPlanCheckoutRequest {
    payment_method: String,
    payment_provider: String,
    payment_channel: Option<String>,
}

fn billing_storage_unavailable_response() -> Response<Body> {
    build_auth_error_response(
        http::StatusCode::SERVICE_UNAVAILABLE,
        BILLING_STORAGE_UNAVAILABLE_DETAIL,
        false,
    )
}

fn normalize_optional_checkout_string(value: Option<String>, max_len: usize) -> Option<String> {
    value
        .map(|value| value.trim().to_ascii_lowercase())
        .filter(|value| !value.is_empty() && value.chars().count() <= max_len)
}

fn normalize_checkout_request(
    payload: BillingPlanCheckoutRequest,
) -> Result<NormalizedBillingPlanCheckoutRequest, &'static str> {
    let payment_provider = normalize_optional_checkout_string(payload.payment_provider, 30)
        .or_else(|| normalize_optional_checkout_string(payload.payment_method.clone(), 30))
        .unwrap_or_else(|| "epay".to_string());
    if !matches!(
        payment_provider.as_str(),
        "epay" | "alipay" | "wxpay" | "stripe"
    ) {
        return Err("unsupported payment_provider");
    }
    let payment_method = normalize_optional_checkout_string(payload.payment_method, 30)
        .unwrap_or_else(|| payment_provider.clone());
    let payment_channel = normalize_optional_checkout_string(payload.payment_channel, 30)
        .or_else(|| (payment_method != "epay").then_some(payment_method.clone()));
    Ok(NormalizedBillingPlanCheckoutRequest {
        payment_method: payment_provider.clone(),
        payment_provider,
        payment_channel,
    })
}

fn plan_id_from_checkout_path(path: &str) -> Option<String> {
    let trimmed = path.trim_end_matches('/');
    let rest = trimmed.strip_prefix("/api/billing/plans/")?;
    let plan_id = rest.strip_suffix("/checkout")?.trim_matches('/');
    if plan_id.is_empty() || plan_id.contains('/') {
        None
    } else {
        Some(plan_id.to_string())
    }
}

fn billing_order_no(now: chrono::DateTime<chrono::Utc>) -> String {
    format!(
        "pp_{}_{}",
        now.format("%Y%m%d%H%M%S%6f"),
        &Uuid::new_v4().simple().to_string()[..12]
    )
}

fn billing_plan_payload(
    record: &aether_data_contracts::repository::billing::BillingPlanRecord,
) -> serde_json::Value {
    json!({
        "id": record.id,
        "title": record.title,
        "description": record.description,
        "price_amount": record.price_amount,
        "price_currency": record.price_currency,
        "duration_unit": record.duration_unit,
        "duration_value": record.duration_value,
        "enabled": record.enabled,
        "sort_order": record.sort_order,
        "max_active_per_user": record.max_active_per_user,
        "purchase_limit_scope": record.purchase_limit_scope,
        "entitlements": record.entitlements_json,
        "created_at": record.created_at_unix_secs,
        "updated_at": record.updated_at_unix_secs,
    })
}

fn billing_plan_snapshot(
    record: &aether_data_contracts::repository::billing::BillingPlanRecord,
) -> serde_json::Value {
    json!({
        "id": record.id,
        "title": record.title,
        "description": record.description,
        "price_amount": record.price_amount,
        "price_currency": record.price_currency,
        "duration_unit": record.duration_unit,
        "duration_value": record.duration_value,
        "max_active_per_user": record.max_active_per_user,
        "purchase_limit_scope": record.purchase_limit_scope,
        "entitlements": record.entitlements_json,
    })
}

fn plan_has_package_rights(
    record: &aether_data_contracts::repository::billing::BillingPlanRecord,
) -> bool {
    record.entitlements_json.as_array().is_some_and(|items| {
        items.iter().any(|item| {
            matches!(
                item.get("type").and_then(|value| value.as_str()),
                Some("daily_quota" | "membership_group")
            )
        })
    })
}

fn payment_order_payload(
    record: &aether_data::repository::wallet::StoredAdminPaymentOrder,
    plan: &aether_data_contracts::repository::billing::BillingPlanRecord,
) -> serde_json::Value {
    json!({
        "id": record.id,
        "order_no": record.order_no,
        "wallet_id": record.wallet_id,
        "user_id": record.user_id,
        "amount_usd": record.amount_usd,
        "pay_amount": record.pay_amount,
        "pay_currency": record.pay_currency,
        "exchange_rate": record.exchange_rate,
        "payment_method": record.payment_method,
        "gateway_order_id": record.gateway_order_id,
        "gateway_response": sanitize_wallet_gateway_response(record.gateway_response.clone()),
        "status": record.status,
        "order_kind": "plan_purchase",
        "product_id": plan.id,
        "product": billing_plan_payload(plan),
        "created_at": unix_secs_to_rfc3339(record.created_at_unix_ms),
        "paid_at": record.paid_at_unix_secs.and_then(unix_secs_to_rfc3339),
        "credited_at": record.credited_at_unix_secs.and_then(unix_secs_to_rfc3339),
        "expires_at": record.expires_at_unix_secs.and_then(unix_secs_to_rfc3339),
    })
}

fn entitlement_payload(
    record: &aether_data_contracts::repository::billing::UserPlanEntitlementRecord,
) -> serde_json::Value {
    json!({
        "id": record.id,
        "user_id": record.user_id,
        "plan_id": record.plan_id,
        "payment_order_id": record.payment_order_id,
        "status": record.status,
        "starts_at": unix_secs_to_rfc3339(record.starts_at_unix_secs),
        "expires_at": unix_secs_to_rfc3339(record.expires_at_unix_secs),
        "entitlements": record.entitlements_snapshot,
        "created_at": unix_secs_to_rfc3339(record.created_at_unix_secs),
        "updated_at": unix_secs_to_rfc3339(record.updated_at_unix_secs),
    })
}

fn compute_plan_payment_amounts(
    plan: &aether_data_contracts::repository::billing::BillingPlanRecord,
    pay_currency: &str,
    usd_exchange_rate: f64,
) -> Result<(f64, f64), &'static str> {
    if !plan.price_amount.is_finite() || plan.price_amount <= 0.0 || usd_exchange_rate <= 0.0 {
        return Err("套餐价格配置无效");
    }
    if plan.price_currency.eq_ignore_ascii_case(pay_currency) {
        let amount_usd =
            (plan.price_amount / usd_exchange_rate * 100_000_000.0).round() / 100_000_000.0;
        let pay_amount = (plan.price_amount * 100.0).round() / 100.0;
        return Ok((amount_usd, pay_amount));
    }
    if plan.price_currency.eq_ignore_ascii_case("USD") {
        let amount_usd = (plan.price_amount * 100_000_000.0).round() / 100_000_000.0;
        let pay_amount = (plan.price_amount * usd_exchange_rate * 100.0).round() / 100.0;
        return Ok((amount_usd, pay_amount));
    }
    Err("套餐币种与支付网关币种不匹配")
}

pub(super) async fn handle_billing_plans_list(state: &AppState) -> Response<Body> {
    let plans = match state.list_billing_plans(false).await {
        Ok(Some(value)) => value,
        Ok(None) => return billing_storage_unavailable_response(),
        Err(err) => {
            return build_auth_error_response(
                http::StatusCode::INTERNAL_SERVER_ERROR,
                format!("billing plan lookup failed: {err:?}"),
                false,
            )
        }
    };
    let items = plans
        .iter()
        .filter(|plan| plan_has_package_rights(plan))
        .map(billing_plan_payload)
        .collect::<Vec<_>>();
    Json(json!({"items": items, "total": items.len()})).into_response()
}

pub(super) async fn handle_billing_entitlements(
    state: &AppState,
    request_context: &GatewayPublicRequestContext,
    headers: &http::HeaderMap,
) -> Response<Body> {
    let auth = match resolve_authenticated_local_user(state, request_context, headers).await {
        Ok(value) => value,
        Err(response) => return response,
    };
    let entitlements = match state.list_user_plan_entitlements(&auth.user.id).await {
        Ok(Some(value)) => value,
        Ok(None) => return billing_storage_unavailable_response(),
        Err(err) => {
            return build_auth_error_response(
                http::StatusCode::INTERNAL_SERVER_ERROR,
                format!("billing entitlement lookup failed: {err:?}"),
                false,
            )
        }
    };
    let now = Utc::now().timestamp().max(0) as u64;
    let items = entitlements
        .iter()
        .map(|record| {
            let mut payload = entitlement_payload(record);
            payload["active"] = json!(
                record.status == "active"
                    && record.starts_at_unix_secs <= now
                    && record.expires_at_unix_secs > now
            );
            payload
        })
        .collect::<Vec<_>>();
    Json(json!({"items": items, "total": items.len()})).into_response()
}

pub(super) async fn handle_billing_plan_checkout(
    state: &AppState,
    request_context: &GatewayPublicRequestContext,
    headers: &http::HeaderMap,
    request_body: Option<&Bytes>,
) -> Response<Body> {
    let auth = match resolve_authenticated_local_user(state, request_context, headers).await {
        Ok(value) => value,
        Err(response) => return response,
    };
    let Some(plan_id) = plan_id_from_checkout_path(&request_context.request_path) else {
        return build_auth_error_response(http::StatusCode::BAD_REQUEST, "缺少套餐ID", false);
    };
    let payload = match request_body {
        Some(body) if !body.is_empty() => {
            match serde_json::from_slice::<BillingPlanCheckoutRequest>(body) {
                Ok(value) => value,
                Err(_) => {
                    return build_auth_error_response(
                        http::StatusCode::BAD_REQUEST,
                        "输入验证失败",
                        false,
                    )
                }
            }
        }
        _ => BillingPlanCheckoutRequest::default(),
    };
    let checkout_request = match normalize_checkout_request(payload) {
        Ok(value) => value,
        Err(detail) => {
            return build_auth_error_response(http::StatusCode::BAD_REQUEST, detail, false)
        }
    };

    let plan = match state.find_billing_plan(&plan_id).await {
        Ok(Some(value)) if value.enabled => value,
        Ok(Some(_)) => {
            return build_auth_error_response(http::StatusCode::BAD_REQUEST, "套餐已下架", false)
        }
        Ok(None) => {
            return build_auth_error_response(http::StatusCode::NOT_FOUND, "套餐不存在", false)
        }
        Err(err) => {
            return build_auth_error_response(
                http::StatusCode::INTERNAL_SERVER_ERROR,
                format!("billing plan lookup failed: {err:?}"),
                false,
            )
        }
    };
    if !plan_has_package_rights(&plan) {
        return build_auth_error_response(
            http::StatusCode::BAD_REQUEST,
            "余额包已移除，请使用钱包充值功能",
            false,
        );
    }
    match state
        .find_pending_plan_purchase_order_by_user_id(&auth.user.id, &plan.id)
        .await
    {
        Ok(Some(order)) => {
            return build_auth_json_response(
                http::StatusCode::OK,
                json!({
                    "order": payment_order_payload(&order, &plan),
                    "payment_instructions": sanitize_wallet_gateway_response(
                        order.gateway_response.clone()
                    ),
                    "reused_pending_order": true,
                }),
                None,
            )
        }
        Ok(None) => {}
        Err(err) => {
            return build_auth_error_response(
                http::StatusCode::INTERNAL_SERVER_ERROR,
                format!("pending billing checkout lookup failed: {err:?}"),
                false,
            )
        }
    }
    let now = Utc::now();
    let order_no = billing_order_no(now);
    let expires_at = now + chrono::Duration::minutes(30);
    let requested_provider = checkout_request.payment_provider.as_str();
    let payment_method = checkout_request.payment_method.clone();
    let payment_channel =
        checkout_request
            .payment_channel
            .clone()
            .or_else(|| match requested_provider {
                "alipay" => Some("alipay".to_string()),
                "wxpay" => Some("native".to_string()),
                "stripe" => Some("card".to_string()),
                _ => None,
            });
    if requested_provider == "epay" {
        let config = match load_epay_config(state).await {
            Ok(value) => value,
            Err(detail) => {
                return build_auth_error_response(http::StatusCode::BAD_REQUEST, detail, false)
            }
        };
        let payment_channel =
            match resolve_epay_channel(&config, checkout_request.payment_channel.as_deref()) {
                Ok(value) => value,
                Err(detail) => {
                    return build_auth_error_response(http::StatusCode::BAD_REQUEST, detail, false);
                }
            };
        let payment_channel_id = payment_channel.channel.clone();
        let (amount_usd, pay_amount) = match compute_plan_payment_amounts(
            &plan,
            &config.pay_currency,
            config.usd_exchange_rate,
        ) {
            Ok(value) => value,
            Err(detail) => {
                return build_auth_error_response(http::StatusCode::BAD_REQUEST, detail, false)
            }
        };
        let Some(callback_base_url) = epay_callback_base_url(
            config.callback_base_url.as_deref(),
            headers,
            request_context,
        ) else {
            return build_auth_error_response(
                http::StatusCode::BAD_REQUEST,
                "epay callback_base_url is required",
                false,
            );
        };
        let checkout = build_epay_checkout_url(
            &config,
            &EpayCheckoutInput {
                order_no: order_no.clone(),
                channel: payment_channel_id.clone(),
                subject: plan.title.clone(),
                pay_amount,
                notify_url: format!("{callback_base_url}/api/payment/epay/notify"),
                return_url: format!("{callback_base_url}/api/payment/epay/return"),
            },
        );
        let outcome = match state
            .create_plan_purchase_order(
                aether_data::repository::wallet::CreatePlanPurchaseOrderInput {
                    preferred_wallet_id: None,
                    user_id: auth.user.id.clone(),
                    amount_usd,
                    pay_amount,
                    pay_currency: config.pay_currency.clone(),
                    exchange_rate: config.usd_exchange_rate,
                    payment_method: payment_method.clone(),
                    payment_provider: Some(checkout_request.payment_provider.clone()),
                    payment_channel: Some(payment_channel_id),
                    gateway_order_id: order_no.clone(),
                    gateway_response: checkout.clone(),
                    order_no: order_no.clone(),
                    product_id: plan.id.clone(),
                    product_snapshot: billing_plan_snapshot(&plan),
                    expires_at_unix_secs: expires_at.timestamp().max(0) as u64,
                },
            )
            .await
        {
            Ok(Some(value)) => value,
            Ok(None) => return billing_storage_unavailable_response(),
            Err(err) => {
                return build_auth_error_response(
                    http::StatusCode::INTERNAL_SERVER_ERROR,
                    format!("billing checkout create failed: {err:?}"),
                    false,
                )
            }
        };
        let order = match outcome {
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::Created(order) => {
                payment_order_payload(&order, &plan)
            }
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::WalletInactive => {
                return build_auth_error_response(
                    http::StatusCode::BAD_REQUEST,
                    "wallet is not active",
                    false,
                )
            }
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::ActivePlanLimitReached => {
                return build_auth_error_response(
                    http::StatusCode::CONFLICT,
                    "套餐购买限制已达到上限",
                    false,
                )
            }
        };
        build_auth_json_response(
            http::StatusCode::OK,
            json!({
                "order": order,
                "payment_instructions": sanitize_wallet_gateway_response(Some(checkout)),
            }),
            None,
        )
    } else {
        let (payment_channel, display_name, pay_currency, usd_exchange_rate, callback_base_url) = {
            let record = match state.find_payment_gateway_config(requested_provider).await {
                Ok(Some(value)) if value.enabled && value.merchant_key_encrypted.is_some() => value,
                Ok(Some(_)) | Ok(None) => {
                    return build_auth_error_response(
                        http::StatusCode::BAD_REQUEST,
                        "支付网关未启用或密钥未配置",
                        false,
                    )
                }
                Err(err) => {
                    return build_auth_error_response(
                        http::StatusCode::INTERNAL_SERVER_ERROR,
                        format!("payment gateway lookup failed: {err:?}"),
                        false,
                    )
                }
            };
            let payment_channel =
                payment_channel
                    .clone()
                    .unwrap_or_else(|| match requested_provider {
                        "alipay" => "alipay".to_string(),
                        "wxpay" => "native".to_string(),
                        "stripe" => "card".to_string(),
                        _ => "alipay".to_string(),
                    });
            let display_name = match requested_provider {
                "alipay" => "支付宝官方".to_string(),
                "wxpay" => match payment_channel.as_str() {
                    "h5" => "微信 H5".to_string(),
                    "jsapi" => "微信 JSAPI".to_string(),
                    _ => "微信 Native".to_string(),
                },
                "stripe" => match payment_channel.as_str() {
                    "alipay" => "Stripe Alipay".to_string(),
                    "wechat_pay" => "Stripe WeChat Pay".to_string(),
                    "link" => "Stripe Link".to_string(),
                    _ => "Stripe Card".to_string(),
                },
                _ => "支付".to_string(),
            };
            (
                payment_channel,
                display_name,
                record.pay_currency,
                record.usd_exchange_rate,
                record.callback_base_url,
            )
        };
        let (amount_usd, pay_amount) =
            match compute_plan_payment_amounts(&plan, &pay_currency, usd_exchange_rate) {
                Ok(value) => value,
                Err(detail) => {
                    return build_auth_error_response(http::StatusCode::BAD_REQUEST, detail, false)
                }
            };
        let Some(callback_base_url) =
            epay_callback_base_url(callback_base_url.as_deref(), headers, request_context)
        else {
            return build_auth_error_response(
                http::StatusCode::BAD_REQUEST,
                "支付网关 callback_base_url is required",
                false,
            );
        };
        let direct_input = DirectPaymentCheckoutInput {
            payment_channel: payment_channel.clone(),
            display_name,
            order_no: order_no.clone(),
            subject: plan.title.clone(),
            pay_amount,
            pay_currency: pay_currency.clone(),
            notify_url: format!("{callback_base_url}/api/payment/{requested_provider}/notify"),
            return_url: Some(format!("{callback_base_url}/dashboard/billing")),
            client_ip: direct_payment_client_ip(headers),
            expires_at,
        };
        let checkout = match requested_provider {
            "alipay" => match create_alipay_direct_checkout(state, &direct_input).await {
                Ok(value) => value,
                Err(detail) => {
                    return build_auth_error_response(http::StatusCode::BAD_GATEWAY, detail, false)
                }
            },
            "wxpay" => match create_wxpay_direct_checkout(state, &direct_input).await {
                Ok(value) => value,
                Err(detail) => {
                    return build_auth_error_response(http::StatusCode::BAD_GATEWAY, detail, false)
                }
            },
            "stripe" => match create_stripe_direct_checkout(state, &direct_input).await {
                Ok(value) => value,
                Err(detail) => {
                    return build_auth_error_response(http::StatusCode::BAD_GATEWAY, detail, false)
                }
            },
            _ => {
                return build_auth_error_response(
                    http::StatusCode::BAD_REQUEST,
                    "unsupported payment provider",
                    false,
                )
            }
        };
        let outcome = match state
            .create_plan_purchase_order(
                aether_data::repository::wallet::CreatePlanPurchaseOrderInput {
                    preferred_wallet_id: None,
                    user_id: auth.user.id.clone(),
                    amount_usd,
                    pay_amount,
                    pay_currency: pay_currency.clone(),
                    exchange_rate: usd_exchange_rate,
                    payment_method,
                    payment_provider: Some(requested_provider.to_string()),
                    payment_channel: Some(payment_channel.clone()),
                    gateway_order_id: checkout
                        .get("gateway_order_id")
                        .and_then(Value::as_str)
                        .unwrap_or(&order_no)
                        .to_string(),
                    gateway_response: checkout.clone(),
                    order_no: order_no.clone(),
                    product_id: plan.id.clone(),
                    product_snapshot: billing_plan_snapshot(&plan),
                    expires_at_unix_secs: expires_at.timestamp().max(0) as u64,
                },
            )
            .await
        {
            Ok(Some(value)) => value,
            Ok(None) => return billing_storage_unavailable_response(),
            Err(err) => {
                return build_auth_error_response(
                    http::StatusCode::INTERNAL_SERVER_ERROR,
                    format!("billing checkout create failed: {err:?}"),
                    false,
                )
            }
        };
        let order = match outcome {
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::Created(order) => {
                payment_order_payload(&order, &plan)
            }
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::WalletInactive => {
                return build_auth_error_response(
                    http::StatusCode::BAD_REQUEST,
                    "wallet is not active",
                    false,
                )
            }
            aether_data::repository::wallet::CreatePlanPurchaseOrderOutcome::ActivePlanLimitReached => {
                return build_auth_error_response(
                    http::StatusCode::CONFLICT,
                    "套餐购买限制已达到上限",
                    false,
                )
            }
        };
        build_auth_json_response(
            http::StatusCode::OK,
            json!({
                "order": order,
                "payment_instructions": sanitize_wallet_gateway_response(Some(checkout)),
            }),
            None,
        )
    }
}

pub(super) async fn maybe_build_local_billing_response(
    state: &AppState,
    request_context: &GatewayPublicRequestContext,
    headers: &http::HeaderMap,
    request_body: Option<&Bytes>,
) -> Option<Response<Body>> {
    let decision = request_context.control_decision.as_ref()?;
    if decision.route_family.as_deref() != Some("billing") {
        return None;
    }
    match decision.route_kind.as_deref() {
        Some("plans") if request_context.request_path == "/api/billing/plans" => {
            Some(handle_billing_plans_list(state).await)
        }
        Some("plan_checkout") => {
            Some(handle_billing_plan_checkout(state, request_context, headers, request_body).await)
        }
        Some("entitlements") if request_context.request_path == "/api/billing/entitlements" => {
            Some(handle_billing_entitlements(state, request_context, headers).await)
        }
        _ => None,
    }
}
